File: logger/server/logger.go
package logger

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"time"
)

var ()

var (
	logFile    *os.File
	logger     *log.Logger
	logDir     string
	initOnce   sync.Once
	parentDir  string
)

func InitLogger(componentName string) {
	initOnce.Do(func() {
		// Get the absolute path to the 'Assignment' root directory dynamically.
		executablePath, err := os.Executable()
		if err != nil {
			log.Fatalf("Failed to get executable path: %v", err)
		}
		// Assuming 'Assignment' is the root directory
		parentDir = filepath.Dir(filepath.Dir(executablePath)) 

		// Create logs directory if it doesn't exist
		logsRoot := filepath.Join(parentDir, "logs")
		if err := os.MkdirAll(logsRoot, 0755); err != nil {
			log.Fatalf("Failed to create logs root directory: %v", err)
		}

		// Create a new log directory for the current run
		currentTime := time.Now().Format("02-01-2006_15:04") // DD-MM-YYYY_HH:MM
		logDir = filepath.Join(logsRoot, currentTime)
		if err := os.MkdirAll(logDir, 0755); err != nil {
			log.Fatalf("Failed to create runtime log directory: %v", err)
		}

		// Set up log file
		logFilePath := filepath.Join(logDir, fmt.Sprintf("%s.log", componentName))
		file, err := os.OpenFile(logFilePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
		if err != nil {
			log.Fatalf("Failed to open log file: %v", err)
		}
		logFile = file

		logger = log.New(logFile, "", 0) // No default flags, custom formatting below
	})
}

func CleanupLogDir() {
	if logFile != nil {
		logFile.Close()
	}

	if logDir != "" {
		// Only remove the directory if it's empty or contains only our log file
		_ = os.RemoveAll(logDir)
	}
}

func Logf(format string, v ...interface{}) {
	if logger == nil {
		log.Printf("Logger not initialized: "+format, v...)
		return
	}

	_, file, line, ok := runtime.Caller(1)
	if !ok {
		file = "???"
		line = 0
	}

	fileName := filepath.Base(file)
	
	message := fmt.Sprintf(format, v...)
	logEntry := fmt.Sprintf("%s[%d]: \"%s\"", fileName, line, strings.TrimSpace(message))

	logger.Println(logEntry)
}

File: logger/client/logger.rs
use std::fs::{self, OpenOptions};
use std::io::{self, Write};
use std::path::{PathBuf};
use chrono::Local;
use once_cell::sync::OnceCell;

pub static LOG_DIR: OnceCell<PathBuf> = OnceCell::new();
pub static LOG_FILE: OnceCell<std::fs::File> = OnceCell::new();

pub fn init_logger(component_name: &str) -> io::Result<()> {
    LOG_DIR.get_or_try_init(|| {
        let assignment_root = std::env::current_dir()?;
        let logs_root = assignment_root.join("logs");

        fs::create_dir_all(&logs_root)?;

        let current_time = Local::now().format("%d-%m-%Y_%H:%M").to_string();
        let log_run_dir = logs_root.join(current_time);
        fs::create_dir_all(&log_run_dir)?;
        
        let log_file_path = log_run_dir.join(format!("{}.log", component_name));
        let file = OpenOptions::new()
            .append(true)
            .create(true)
            .write(true)
            .open(&log_file_path)?;

        LOG_FILE.set(file).map_err(|_| io::Error::new(io::ErrorKind::Other, "Failed to set log file"))?;
        Ok(log_run_dir)
    }).map(|_| ()) // Return an empty Ok once initialized
}

pub fn cleanup_log_dir() {
    if let Some(path) = LOG_DIR.get() {
        let _ = fs::remove_dir_all(path);
    }
}

#[macro_export]
macro_rules! log_entry {
    ($($arg:tt)*) => ({
        use chrono::Local;
        use std::io::Write;
        let msg = format!($($arg)*);
        let file = file!();
        let line = line!();
        let log_line = format!("{}[{}]: \"{}\"\n", file, line, msg);
        
        if let Some(mut file) = $crate::logger::LOG_FILE.get() {
            let _ = file.write_all(log_line.as_bytes());
        } else {
            // Fallback to stderr if logger not initialized
            eprint!("Logger not initialized: {}", log_line);
        }
    });
}

pub use log_entry;

pub fn set_log_file_for_tests(file: std::fs::File) {
    let _ = LOG_FILE.set(file);
}
